README.md
# bock [![NPM version](https://badge.fury.io/js/bock.svg)](https://npmjs.org/package/bock)   [![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/feross/standard)   [![Build Status](https://travis-ci.com/doesdev/bock.svg)](https://travis-ci.com/doesdev/bock)

> Bock is a strong ~~lager~~ logger of ~~German~~ US origin

## install

```sh
$ npm install --save bock
```

## usage

```js
const options = { appName: 'myAwesomeApp', logLevel: 'warn', toFile: true }
const bock = require('bock')(options)
bock.fatal(new Error('something went wrong'))
```

## api

Bock exports a primary function which returns an instance with methods
(`debug`, `info`, `warn`, `fatal`, `setLogLevel`, `close`)
- **options** *(Object - optional)*
  - **appName** *(String - optional - default: 'bock')* Base name to use for log files (`${appName}-${month}-${day}-${year}.json`)
  - **logBase** *(String - optional - default: `__dirname + '/logs'`)* Path for log files
  - **logLevel** *(String - optional - default: 'debug')* Threshold for logging (i.e. if warn it won't log debug or info errors)
  - **newline** *(Boolean - optional - default: `true`)* Use newline delimited JSON (highly recommended for performance reasons)
  - **toConsole** *(Boolean - optional - default: true)* Should log to console
  - **toFile** *(Boolean - optional - default: true)* Should log to file
  - **whitelist** *(Array - optional)* List of ignored error types / messages
  - **track** *(Boolean - optional - default: false)* Should log methods return
  a `Promise` indicating the log was written

Each of the four methods (`debug`, `info`, `warn`, `fatal`) accept an `Error`
object, a string, or a function that returns one of those two things. They also
accept a second argument (`transform`) which is a function that will accept as
its only argument the final text output to file and/or console, and what it
returns will the text value to be logged. This is primarily useful for
cleansing sensitive information from the output.

Where `const logger = bock([opts])` (i.e. on an instance) the following methods
are also available

`logger.setLogLevel(string logLevel)` Changes logLevel setting on instance

`logger.close()` Close forked process used for file writing

Additionally, you can choose to use the `bock.cached([opts])` which will return
the last instance instantiated using the `cached` method or create a new one
with the passed opts (and cache it).

## license

MIT © [Andrew Carpenter](https://github.com/doesdev)


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

#### [4.1.0](https://github.com/doesdev/bock/compare/4.0.1...4.1.0)

> 22 September 2021

- Add `transform` option to all log methods
- Update `mvt`

#### [4.0.1](https://github.com/doesdev/bock/compare/4.0.0...4.0.1)

> 12 November 2020

- Pass empty execArgv to fork to prevent passing unwanted args down

#### [4.0.0](https://github.com/doesdev/bock/compare/3.0.4...4.0.0)

> 16 May 2020

- Use streams for writing instead of appendFileSync
- Update Node Engine to >= 8.0.0

#### [3.0.4](https://github.com/doesdev/bock/compare/3.0.3...3.0.4)

> 3 February 2020

- Store logLevel as text rather than int

#### [3.0.3](https://github.com/doesdev/bock/compare/3.0.2...3.0.3)

> 2 February 2020

- Add `track` option, allowing for awaiting log writing (for tests mainly)

#### [3.0.2](https://github.com/doesdev/bock/compare/3.0.1...3.0.2)

> 1 February 2020

- Retain functionality but don't use OO style
- Fixes inability to use destructured
- Make `cached` more intelligent

#### [3.0.1](https://github.com/doesdev/bock/compare/3.0.0...3.0.1)

> 1 February 2020

- Only cache an instance created with `cached`

#### [3.0.0](https://github.com/doesdev/bock/compare/2.2.0...3.0.0)

> 1 February 2020

- Rewrite OO-esque for no good reason and bunch of other pointless garbage
- Allow functions as message
- Add cached getter

#### [2.2.0](https://github.com/doesdev/bock/compare/2.1.4...2.2.0)

> 30 December 2019

- On `.close` disconnect child rather than kill, fixes #9
- Update `app-root-path` dep
- Fix standard style issues

#### [2.1.4](https://github.com/doesdev/bock/compare/2.1.3...2.1.4)

> 14 July 2019

- Simplify devDependencies
- Fix standard issues
- Add Travis config
- Fix a couple `close` issues

#### [2.1.3](https://github.com/doesdev/bock/compare/2.1.2...2.1.3)

> 1 June 2019

- Update dependencies

#### [2.1.2](https://github.com/doesdev/bock/compare/2.1.1...2.1.2)

> 10 August 2018

- Add close option (fixes #7), changelog, update dep [`#7`](https://github.com/doesdev/bock/issues/7)

#### [2.1.1](https://github.com/doesdev/bock/compare/2.1.0...2.1.1)

> 10 May 2018

- Add setLogLevel, 2.1.1 [`e2ea3d9`](https://github.com/doesdev/bock/commit/e2ea3d96f7c3ad821e8214fea880db5e454019b0)

#### [2.1.0](https://github.com/doesdev/bock/compare/2.0.0...2.1.0)

> 30 November 2017

- Fix #8, console logging in debug [`#8`](https://github.com/doesdev/bock/issues/8)

#### 2.0.0

> 29 October 2017

- Breaking 2.0.0, fixes #4 and #5, newline delimited [`#4`](https://github.com/doesdev/bock/issues/4)
- Whitelist message, fixes #2, v1.0.5 [`#2`](https://github.com/doesdev/bock/issues/2)
- Fix #1, use app-root-path for default logBase [`#1`](https://github.com/doesdev/bock/issues/1)
- Add deps badge [`f5437fa`](https://github.com/doesdev/bock/commit/f5437fa4390162dff012d477bdbc8993d368826f)
- Fix that last thing [`f0bc6b0`](https://github.com/doesdev/bock/commit/f0bc6b0c8249008bc67ed60040e2e59803241472)
- Revert naming convention to original [`ddd66a4`](https://github.com/doesdev/bock/commit/ddd66a44384fda25eba751a4740d2a046935bad0)
- Use hoy for cached today [`62d4cbf`](https://github.com/doesdev/bock/commit/62d4cbfa3e9ba800cc8450a99b91a8f4310de4a7)
- Fix bad commit [`095c9de`](https://github.com/doesdev/bock/commit/095c9dee08e67a72ba7ef4a012f2264b9f7bb0d1)
- Bump to v1.0.1 [`bbe74fd`](https://github.com/doesdev/bock/commit/bbe74fde95c49ba855196eae3181f7d51b853f92)
- Initial commit [`5c1e3c4`](https://github.com/doesdev/bock/commit/5c1e3c46d8defcd3c1cc6a300087d8f07726273b)


test.js
'use strict'

const test = require('mvt')
const fs = require('fs')
const path = require('path')
const bock = require('./index')
const getOpts = (override = {}) => Object.assign({}, {
  appName: 'bocktest',
  logBase: path.join(__dirname, '_logs'),
  logLevel: 'debug',
  toConsole: false,
  newline: false,
  track: true
}, override)

const clearAll = (logger, clearDir) => {
  const logBase = getOpts().logBase
  try {
    const logFiles = fs.readdirSync(logBase).map((f) => path.join(logBase, f))
    logFiles.forEach((f) => { try { fs.unlinkSync(f) } catch (ex) {} })
    if (clearDir) fs.rmdirSync(getOpts().logBase)
  } catch (ex) {}

  if (logger) logger.close()
}

test.before(() => clearAll(null, true))

test.after(() => clearAll(null, true))

test('proper JSON array', async (assert) => {
  const opts = getOpts({ appName: 'json-array' })
  const logger = bock(opts)

  await logger.debug(new Error('debug'))
  await logger.info(new Error('info'))
  await logger.warn(new Error('warn'))
  await logger.fatal(new Error('fatal'))

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const log = require(logFile)

  assert.is(log[0].level, 'debug')
  assert.is(log[1].level, 'info')
  assert.is(log[2].level, 'warn')
  assert.is(log[3].level, 'fatal')

  clearAll(logger)
})

test('newline delimited JSON', async (assert) => {
  const opts = getOpts({ appName: 'newline', newline: true })
  const logger = bock(opts)

  await logger.debug(new Error('debug'))
  await logger.info(new Error('info'))
  await logger.warn(new Error('warn'))
  await logger.fatal(new Error('fatal'))

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const nld = fs.readFileSync(logFile, 'utf8').trim().split('\n').join(',')
  const log = JSON.parse(`[${nld}]`)

  assert.is(log[0].level, 'debug')
  assert.is(log[1].level, 'info')
  assert.is(log[2].level, 'warn')
  assert.is(log[3].level, 'fatal')

  clearAll(logger)
})

test('circular reference', async (assert) => {
  const opts = getOpts({ appName: 'circular' })
  const logger = bock(opts)
  const err = new Error('debug')
  const meta = { error: err }

  err.meta = meta

  assert.notThrows(() => logger.debug(err))

  clearAll(logger)
})

test('console output is as expected', async (assert) => {
  const logger = bock({ appName: 'a', toConsole: true, toFile: false })
  let val, old

  // debug
  old = console.log
  console.log = (v) => { val = v }
  logger.debug(new Error('debug'))
  assert.truthy(val.match(/^DEBUG/))
  console.log = old

  // info
  old = console.info
  console.info = (v) => { val = v }
  logger.info(new Error('info'))
  assert.truthy(val.match(/^INFO/))
  console.info = old

  // warn
  old = console.warn
  console.warn = (v) => { val = v }
  logger.warn(new Error('warn'))
  assert.truthy(val.match(/^WARN/))
  console.warn = old

  // fatal
  old = console.error
  console.error = (v) => { val = v }
  logger.fatal(new Error('fatal'))
  assert.truthy(val.match(/^FATAL/))
  console.error = old

  // change logLevel to fatal
  logger.setLogLevel('fatal')

  // run debug scenario, but we expect val to be unchanged from FATAL
  old = console.log
  console.log = (v) => { val = v }
  logger.debug(new Error('debug'))
  assert.truthy(val.match(/^FATAL/))
  console.log = old

  clearAll(logger)
})

test('function as error', async (assert) => {
  const opts = getOpts({ appName: 'function' })
  const logger = bock(opts)

  await logger.fatal(() => new Error('fatal'))

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const log = require(logFile)

  assert.is(log[0].level, 'fatal')

  clearAll(logger)
})

test('transform error text', async (assert) => {
  const opts = getOpts({ appName: 'transform' })
  const logger = bock(opts)

  await logger.fatal(new Error('fatal'), (t) => t.replace('ata', 'ootbal'))

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const log = require(logFile)

  assert.is(log[0].level, 'football')

  clearAll(logger)
})

test('whitelist', async (assert) => {
  const opts = getOpts({ appName: 'whitelist', whitelist: ['info'] })
  const logger = bock(opts)

  await logger.debug(new Error('debug'))
  await logger.info(new Error('info'))
  await logger.warn(new Error('warn'))
  await logger.fatal(new Error('fatal'))

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const log = require(logFile)

  assert.is(log[0].level, 'debug')
  assert.is(log[1].level, 'warn')
  assert.is(log[2].level, 'fatal')

  clearAll(logger)
})

test('cached returns last instance', async (assert) => {
  const loggerInit = bock.cached(getOpts({ appName: 'cached' }))
  const loggerCached = bock.cached()
  assert.is(loggerInit, loggerCached)
  loggerCached.close()
})

test('cached re-inits if cached needs opts', async (assert) => {
  const loggerNeedsOpts = bock.cached()
  const loggerReInit = bock.cached(getOpts({ appName: 'cache-reinit' }))
  const loggerCached = bock.cached()
  assert.not(loggerNeedsOpts, loggerReInit)
  assert.not(loggerNeedsOpts, loggerCached)
  assert.is(loggerReInit, loggerCached)
  loggerCached.close()
})

test('can destructure', async (assert) => {
  const opts = getOpts({ appName: 'destructure' })
  const logger = bock(opts)
  const { debug, info, warn, fatal } = logger

  await debug(new Error('debug'))
  await info(new Error('info'))
  await warn(new Error('warn'))
  await fatal(new Error('fatal'))

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const log = require(logFile)

  assert.is(log[0].level, 'debug')
  assert.is(log[1].level, 'info')
  assert.is(log[2].level, 'warn')
  assert.is(log[3].level, 'fatal')

  clearAll(logger)
})

test('methods return promise with track = true', async (assert) => {
  const logger = bock(getOpts({ appName: 'track-true', track: true }))
  assert.true(logger.debug('debug') instanceof Promise)

  clearAll(logger)
})

test('methods return null with track = false', async (assert) => {
  const logger = bock(getOpts({ appName: 'track-false', track: false }))
  assert.is(logger.debug('debug'), null)

  clearAll(logger)
})

test('stress test with newline = false', async (assert) => {
  const opts = getOpts({ appName: 'stress-array', track: true })
  const logger = bock(opts)
  const runs = 1000
  const start = Date.now()

  await Promise.all([...Array(runs)].map((_, i) => {
    return logger.debug(new Error(`debug: ${i}`))
  }))

  const fin = Date.now()
  const ranFor = fin - start

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const log = require(logFile)

  assert.is(log.length, runs)
  assert.true(ranFor < 30000)

  console.log(`Stress-Array: ${runs} log entries committed in ${ranFor}ms`)

  clearAll(logger)
})

test('stress test with newline = true', async (assert) => {
  const opts = getOpts({ appName: 'stress-newline', track: true, newline: true })
  const logger = bock(opts)
  const runs = 5000
  const start = Date.now()

  await Promise.all([...Array(runs)].map((_, i) => {
    return logger.debug(new Error(`debug: ${i + 1}`))
  }))

  const fin = Date.now()
  const ranFor = fin - start

  const logFile = path.resolve(opts.logBase, fs.readdirSync(opts.logBase)[0])
  const nld = fs.readFileSync(logFile, 'utf8').trim().split('\n').join(',')
  const log = JSON.parse(`[${nld}]`)

  assert.is(log.length, runs)
  assert.is(log[log.length - 1].message, `debug: ${runs}`)
  assert.true(ranFor < 10000)

  console.log(`Stress-Newline: ${runs} log entries committed in ${ranFor}ms`)

  clearAll(logger)
})


index.js
'use strict'

const {
  lvlsInt,
  defaultBase,
  writer,
  randar,
  kId,
  kAppName,
  kLogBase,
  kLogLevel,
  kNewline,
  kToConsole,
  kToFile,
  kWhitelist,
  kTrack,
  mapWhiteList,
  logIt,
  closeForked
} = require('./util')

const instances = {}
let cached

const bock = (id, opts) => {
  const needsOpts = !opts
  opts = opts || {}

  const instOpts = {
    [kAppName]: opts.appName || 'bock',
    [kLogBase]: opts.logBase || defaultBase,
    [kLogLevel]: lvlsInt[opts.logLevel] ? opts.logLevel : 'debug',
    [kNewline]: opts.newline !== false,
    [kToConsole]: opts.toConsole !== false,
    [kToFile]: opts.toFile !== false,
    [kWhitelist]: mapWhiteList(opts.whitelist),
    [kTrack]: opts.track === true
  }

  if (instOpts[kToFile]) {
    try {
      require('fs').mkdirSync(instOpts[kLogBase])
    } catch (e) {}
    writer()
  }

  const bockInstance = {
    [kId]: id,
    get appName () { return instOpts[kAppName] },
    get logLevel () { return instOpts[kLogLevel] },
    set logLevel (ll) {
      instOpts[kLogLevel] = lvlsInt[ll] ? ll : instOpts[kLogLevel] || 'debug'
    },
    setLogLevel (logLevel) { bockInstance.logLevel = logLevel },
    debug: (err, transform) => logIt(err, 'debug', instOpts, transform),
    info: (err, transform) => logIt(err, 'info', instOpts, transform),
    warn: (err, transform) => logIt(err, 'warn', instOpts, transform),
    fatal: (err, transform) => logIt(err, 'fatal', instOpts, transform),
    close () {
      delete instances[id]
      if (cached && cached[kId] === id) cached = undefined
      if (!Object.keys(instances).length) closeForked()
    },
    needsOpts
  }

  return bockInstance
}

const newBock = (opts) => {
  const id = `${(opts || {}).appName || 'bock'}-${Date.now()}-${randar()}`
  return (instances[id] = bock(id, opts))
}

module.exports = newBock

module.exports.cached = (opts) => {
  if (cached && cached.needsOpts && opts) return (cached = newBock(opts))
  return (cached = cached || newBock(opts))
}


writer.js
'use strict'

const fs = require('fs')
const maxAttempts = 10
const prepMsg = (data, msgType, msg) => `${data.id}|${msgType}|${msg || 'done'}`
const queue = []
const stream = {}

const write = () => new Promise((resolve, reject) => {
  const { log, logFilePath: fPath } = queue.shift() || {}
  if (!(log && fPath)) return resolve()

  if (stream.path !== fPath) {
    if (stream.active && stream.active.writable) stream.active.end()
    stream.active = fs.createWriteStream(fPath, { flags: 'a' })
    stream.path = fPath
  }

  stream.active.write(`${log}\n`, () => resolve())
})

const writeIt = async (data, attempts) => {
  try {
    queue.push(data)
    await write()
    if (data.id) process.send(prepMsg(data, 'done'))
  } catch (e) {
    if (attempts < maxAttempts) return writeIt(data, attempts + 1)
    process.send(data.id ? prepMsg(data, 'error', e) : e)
  }
}

const writeItAsAry = (data, attempts) => {
  try {
    var logs = require(data.logFilePath)
  } catch (e) {
    logs = []
  }
  logs.push(JSON.parse(data.log))
  try {
    fs.writeFileSync(data.logFilePath, JSON.stringify(logs))
    if (data.id) process.send(prepMsg(data, 'done'))
  } catch (e) {
    if (attempts < maxAttempts) return writeItAsAry(data, attempts + 1)
    process.send(data.id ? prepMsg(data, 'error', e) : e)
  }
}

process.on('message', (d) => d.newline ? writeIt(d, 0) : writeItAsAry(d, 0))


util.js
'use strict'

const path = require('path')
const hoy = require('hoy')
const appRoot = require('app-root-path').toString()

const lvlsInt = { debug: 1, info: 2, warn: 3, fatal: 4 }
const defaultBase = path.join(appRoot, 'logs')
const errProps = [
  'name',
  'message',
  'fileName',
  'lineNumber',
  'stack',
  'meta'
]

const kId = Symbol('id')
const kAppName = Symbol('appName')
const kLogBase = Symbol('logBase')
const kLogLevel = Symbol('logLevel')
const kNewline = Symbol('newline')
const kToConsole = Symbol('toConsole')
const kToFile = Symbol('toFile')
const kWhitelist = Symbol('whitelist')
const kTrack = Symbol('track')

// globals
const completionCallbacks = {}
let commitLogToFile

const fireCallbackIfNeeded = (id) => {
  const capturedCb = completionCallbacks[id]
  if (!capturedCb) return
  delete completionCallbacks[id]
  capturedCb()
}

const writer = (toWrite, cb) => {
  if (!commitLogToFile || !commitLogToFile.connected) {
    const fork = require('child_process').fork
    const writerWorker = path.join(__dirname, 'writer.js')

    if (commitLogToFile && commitLogToFile.removeAllListeners) {
      commitLogToFile.removeAllListeners('close')
      commitLogToFile.removeAllListeners('message')
      if (commitLogToFile.kill) try { commitLogToFile.kill() } catch (ex) {}
      if (commitLogToFile.unfref) try { commitLogToFile.unfref() } catch (ex) {}
    }

    commitLogToFile = fork(writerWorker, { execArgv: [] })
    commitLogToFile.on('close', () => writer())
    commitLogToFile.on('message', (msg) => {
      const msgTypes = { error: true, done: true }
      const [id, msgType, ...body] = msg.split('|')
      if (!msgTypes[msgType]) return console.error(new Error(msg))
      if (msgType === 'error') console.error(new Error(body.join('|')))
      fireCallbackIfNeeded(id)
    })
  }

  if (!toWrite) return

  if (cb) {
    const id = toWrite.id = `${Date.now()}-${randar()}`
    completionCallbacks[id] = cb
    setTimeout(() => fireCallbackIfNeeded(id), 30000)
  }

  commitLogToFile.send(toWrite)
}

const today = () => {
  const d = hoy()
  return `${d.month}-${d.day}-${d.year}`
}

const initCap = (str) => {
  str = str || ''
  return `${str.charAt(0).toUpperCase()}${str.slice(1)}`
}

const randar = () => {
  return (
    Math.random().toString(36).slice(2) +
    Math.random().toString(36).slice(2)
  ).substring(0, 16)
}

const stringer = (obj) => {
  try {
    return JSON.stringify(obj)
  } catch (ex) {
    const newObj = {}
    try {
      Object.keys(obj).forEach((k) => {
        try {
          newObj[k] = JSON.parse(JSON.stringify(obj[k]))
        } catch (ex) {}
      })
      return JSON.stringify(newObj)
    } catch (ex) {
      return '{"bockError":"JSON.stringify failed for error"}'
    }
  }
}

const mapWhiteList = (wl) => {
  if (!wl) return new Map()
  return new Map((Array.isArray(wl) ? wl : [wl]).map((k) => [k]))
}

const logIt = (err = new Error(), level = 'warn', instance, transform) => {
  const toReturn = instance[kTrack] ? Promise.resolve() : null
  if (lvlsInt[level] < lvlsInt[instance[kLogLevel]]) return toReturn

  if (typeof err === 'function') err = err()

  const appName = instance[kAppName]
  const toFile = instance[kToFile]
  const toConsole = instance[kToConsole]
  const newline = instance[kNewline]
  const name = err.name || err.toString()
  const whitelist = instance[kWhitelist]
  const tranIsFunc = typeof transform === 'function'

  if (whitelist.has(name) || whitelist.has(err.message)) return toReturn

  const log = {}
  log.time = Date.now()
  log.level = level

  if (typeof err === 'string') {
    log.message = err
  } else {
    Object.assign(log, err)
    errProps.forEach((p) => { if (err[p]) log[p] = err[p] })
  }

  if (!log.name && !log.message) log.message = err

  const logProps = Object.getOwnPropertyNames(log)

  let logText = `${level.toUpperCase()}:`
  logText += `\n  Timestamp: ${new Date(log.time).toLocaleString()}`

  logProps.forEach((k) => {
    if (k === 'time' || k === 'level') return
    const val = typeof log[k] === 'object' ? stringer(log[k]) : log[k]
    logText += `\n  ${initCap(k)}: ${val}`
  })

  if (toConsole) {
    const alias = { debug: 'log', fatal: 'error' }
    const logFunc = console[alias[level] || level] || console.log

    if (tranIsFunc) logText = transform(logText)

    logFunc(logText)
  }

  if (toFile) {
    const logBase = instance[kLogBase]
    const logFilePath = path.join(logBase, `${appName}-${today()}.json`)

    if (instance[kTrack]) {
      return new Promise((resolve, reject) => {
        const toWrite = tranIsFunc ? transform(stringer(log)) : stringer(log)
        writer({ logFilePath, log: toWrite, newline }, () => resolve())
      })
    }

    writer({ logFilePath, log: stringer(log), newline })
  }

  return toReturn
}

const closeForked = () => {
  if (!(commitLogToFile && commitLogToFile.connected)) return
  commitLogToFile.removeListener('message', console.error)
  commitLogToFile.removeListener('close', writer)
  commitLogToFile.disconnect()
}

module.exports = {
  lvlsInt,
  defaultBase,
  writer,
  randar,
  kId,
  kAppName,
  kLogBase,
  kLogLevel,
  kNewline,
  kToConsole,
  kToFile,
  kWhitelist,
  kTrack,
  mapWhiteList,
  logIt,
  closeForked
}
